### PR TITLE
feat: `createAzureAdOAuthConfig()` and `createAzureAdb2cOAuthConfig()`

### PR DESCRIPTION
**Pull request by mcgear**
This would address #115 , i believe i have it working. Haven't tested a lot of different scopes, but it does work with the scopes we are using `https://management.core.windows.net//.default` and `openid`.

Here is an example of creating it:
```
createAzureOAuthConfig({
   scope: ["openid"],
 }),
```

You can also change the prefix used on the environment variables from the default `AZURE` to whatever is needed:
```
createAzureOAuthConfig({
   envPrefix: "MY_PREFIX",
   scope: ["openid"],
 }),
```

### CODE REVIEW

**Comment by mcgear** [comment]
Still working on some issues with azure ADB2C support involving some query string params that are needed for that, and proper redirect

**Comment by mcgear** [comment]
Support for Azure and ADB2C are working

**Comment by mcgear** [comment]
To use this  until it gets merged, you can use an import map like this:

```
    "$fresh/oauth": "https://raw.githubusercontent.com/fathym-deno/deno_kv_oauth/main/mod.ts",
```

File: lib/sign_in.ts
Comment (iuioiua) [inline lib/sign_in.ts:1]: These changes are unrelated to the PR. Can you please revert?

File: lib/handle_callback.ts
Comment (iuioiua) [inline lib/handle_callback.ts:1]: These changes are unrelated to the PR. Can you please revert?

File: lib/create_azure_oauth_config.ts [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+ * ```
+ *
+ * This also supports the ability to connect with Azure AD B2C instances using the
+ * proper configurations for AD B2C instances
+ *
+ * @see {@link https://learn.microsoft.com/en-us/entra/identity-platform/v2-protocols-oidc}
+ */
+export function createAzureOAuthConfig(config: {
+  /** Used to set the prefix used to access environment variables; defaults to `AZURE`.*/
+  envPrefix?: string;
```
Comment (iuioiua) [inline lib/create_azure_oauth_config.ts:52]: This option doesn't exist for any of the other platform config creator functions. Can you please remove it? A single, hardcoded environment variable prefix was a deliberate design decision to simplify logic.

File: lib/create_azure_oauth_config.ts [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+ *   scope: "openid",
+ * });
+ * ```
+ *
+ * This also supports the ability to connect with Azure AD B2C instances using the
+ * proper configurations for AD B2C instances
+ *
+ * @see {@link https://learn.microsoft.com/en-us/entra/identity-platform/v2-protocols-oidc}
+ */
+export function createAzureOAuthConfig(config: {
```
Comment (iuioiua) [inline lib/create_azure_oauth_config.ts:50]: This function differs in its logic from other similar functions within this library. For ease of management of use, could you please try following the implementation of other `createXOAuthConfig()` functions in the library?

**Review by iuioiua — CHANGES_REQUESTED** [review]
Sorry for the delay and thanks for the PR! It'll be good to have this functionality added. A few suggestions, mostly around sticking to how other `createXOAuthConfig()` functions are implemented.

File: lib/create_azure_oauth_config.ts (hunk 1 above)
Comment (mcgear) [reply lib/create_azure_oauth_config.ts:52]: Is there anyway we can include this change? A couple of reasons, one, I was going to create a pr to add it to the other ones. When doing the ahead of time builds with the github oauth provider, this fails as github actions do not allow environment variables that start with `GITHUB_`. Two, by design, and azure capabilities, we have to use two separate Azure providers.  9ne for oauth providers, one for oauth and the other for access to azure infrastructure via oath access tokens.

File: lib/sign_in.ts
Comment (mcgear) [reply lib/sign_in.ts:1]: Reverted

File: lib/handle_callback.ts
Comment (mcgear) [reply lib/handle_callback.ts:1]: Reverted

File: lib/create_azure_oauth_config.ts (hunk 1 above)
Comment (mcgear) [reply lib/create_azure_oauth_config.ts:52]: I also tried to keep it inline with other impls by defaulting to the singular `AZURE` suffix to keep inline when envPrefix is omitted.

Comment (iuioiua) [reply lib/create_azure_oauth_config.ts:52]: Let's hardcode the environment variable prefix for now, like for every other provider. We've kept this API's library simple and consistent while working for most use cases. Edge cases can be handled by using custom OAuth configs.

File: lib/create_azure_oauth_config.ts [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+    clientId,
+    clientSecret: getRequiredEnv(getEnvKey("CLIENT_SECRET")),
+    authorizationEndpointUri: `${baseURL}/v2.0/authorize`,
+    tokenUri: `${baseURL}/v2.0/token`,
+    redirectUri: config.redirectUri,
+    defaults: {
+      scope: config.scope,
+    },
+  };
+}
```
Comment (iuioiua) [inline lib/create_azure_oauth_config.ts:97]:
```suggestion
/**
 * Returns the OAuth configuration for Azure.
 *
 * Requires `--allow-env[=AZURE_CLIENT_ID,AZURE_CLIENT_SECRET,AZURE_TENANT_ID]`
 * permissions and environment variables:
 * 1. `AZURE_CLIENT_ID`
 * 2. `AZURE_CLIENT_SECRET`
 * 3. `AZURE_TENANT_ID`
 *
 * @example
 * ```ts
 * import { createAzureOAuthConfig } from "https://deno.land/x/deno_kv_oauth@$VERSION/mod.ts";
 *
 * const oauthConfig = createAzureOAuthConfig({
 *   redirectUri: "http://localhost:8000/callback",
 *   scope: "https://graph.microsoft.com/User.Read",
 * });
 * ```
 *
 * @see {@link https://learn.microsoft.com/en-us/entra/identity-platform/v2-oauth2-auth-code-flow}
 */
export function createAzureOAuthConfig(config: {
  /** @see {@linkcode OAuth2ClientConfig.redirectUri} */
  redirectUri?: string;
  /** @see {@linkcode OAuth2ClientConfig.defaults.scope} */
  scope: string | string[];
}): OAuth2ClientConfig {
  const baseUrl = `https://login.microsoftonline.com/${
    getRequiredEnv("AZURE_TENANT_ID")
  }/oauth2/v2.0`;
  return {
    clientId: getRequiredEnv("AZURE_CLIENT_ID"),
    clientSecret: getRequiredEnv("AZURE_CLIENT_SECRET"),
    authorizationEndpointUri: `${baseUrl}/authorize`,
    tokenUri: `${baseUrl}/token`,
    redirectUri: config.redirectUri,
    defaults: {
      scope: config.scope,
    },
  };
}
```
These changes bring the Azure implementation into line with the other platforms. I've tested it, and it works as expected. I want to clarify that these functions are meant to cater to the majority of use cases, not all use cases. Advanced use cases can take advantage of custom OAuth configs, if needed.

**Review by iuioiua — CHANGES_REQUESTED** [review]
PTAL at my suggested implementation. Can you please also add a test to `lib/create_oauth_config_test.ts`? You may need to add the following to the test:
```ts
// For Azure
Deno.env.set(`${envPrefix}_TENANT_ID`, crypto.randomUUID());
```

Regarding multiple providers using for the same platform, let's open a separate issue to discuss. I feel like such a use might just warrant a custom OAuth configuration.

File: lib/create_azure_oauth_config.ts (hunk 3 above)
Comment (mcgear) [reply lib/create_azure_oauth_config.ts]: With these changes azure ADB2C won't work, which is actually what we need to use, the major issue being the hardcoding of https://login.microsoftonline.com.  I split AD and ADB2C into separate config providers. Though i still think we should include the logic around client ID, otherwise the end user experience is more complicated by having to know to add the client ID as a scope instead of just `openid`.

Comment (mcgear) [reply lib/create_azure_oauth_config.ts]: Let me know if you want me to remove the client ID logic.

Tests were added as well.

**Comment by mcgear** [comment]
> PTAL at my suggested implementation. Can you please also add a test to `lib/create_oauth_config_test.ts`? You may need to add the following to the test:
> 
> ```ts
> // For Azure
> Deno.env.set(`${envPrefix}_TENANT_ID`, crypto.randomUUID());
> ```
> 
> Regarding multiple providers using for the same platform, let's open a separate issue to discuss. I feel like such a use might just warrant a custom OAuth configuration.

With the separation of Azure ADB2C from Azure AD, this would work for us now.

**Comment by iuioiua** [comment]
Apologies for the delay, @mcgear. I don't like to drag PRs on for this long. I'll review this PR by the end of this week.

**Comment by mcgear** [comment]
Sounds good, i'll try to be quick on incorporating any other changes, but i think i got both of the new ones inline with your other comments.  The only real last question is whether or not we keep the ClientID code in or force end users to add it themselves when leveraging openid

**Comment by iuioiua** [comment]
The user should add the client ID if they want to leverage OpenID. Looking at the current implementations, they still include logic that isn't present for other platforms. Again, can you please remove this logic?

**Comment by mcgear** [comment]
Those changes are in

**Review by iuioiua — CHANGES_REQUESTED** [review]
This is looking much better! Can you please provide ensure you've followed [these steps](https://github.com/denoland/deno_kv_oauth/blob/main/.github/CONTRIBUTING.md#adding-a-pre-configured-oauth-client) in this PR so we know the client functions are working?

File: README.md [hunk 4]
```diff
@@ -314,6 +314,7 @@ is set in the following order of precedence:
 The following providers have pre-defined OAuth configurations:
 
 1. [Auth0](https://deno.land/x/deno_kv_oauth/mod.ts?s=createAuth0OAuthConfig)
+1. [Azure](https://deno.land/x/deno_kv_oauth/mod.ts?s=createAzureOAuthConfig)
```
[the remark this answers is not in the document]
Comment (mcgear) [reply README.md]: Fixed the readme

**Comment by mcgear** [comment]
Fixed the RE

> This is looking much better! Can you please provide ensure you've followed [these steps](https://github.com/denoland/deno_kv_oauth/blob/main/.github/CONTRIBUTING.md#adding-a-pre-configured-oauth-client) in this PR so we know the client functions are working?

2 and 3 from that link are taken care of, i'm not sure what to do for 1 as the link on contributing is broken. I am using both configurations in a deployment, and they are working. Let me know what sort of screenshot i should be providing.

File: lib/create_azure_adb2c_oauth_config.ts [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+ *
+ * const oauthConfig = createAzureADB2COAuthConfig({
+ *   redirectUri: "http://localhost:8000/callback",
+ *   scope: ["openid", Deno.env.get("AZURE_ADB2C_CLIENT_ID")!]
+ * });
+ * ```
+ *
+ * @see {@link https://learn.microsoft.com/en-us/entra/identity-platform/v2-protocols-oidc}
+ */
+export function createAzureADB2COAuthConfig(config: {
```
Comment (iuioiua) [inline lib/create_azure_adb2c_oauth_config.ts:28]:
```suggestion
export function createAzureAdb2cOAuthConfig(config: {
```
Ditto

File: lib/create_azure_ad_oauth_config.ts [hunk 6]
```diff
@@ -0,0 +1,48 @@
...
+ *
+ * const oauthConfig = createAzureADOAuthConfig({
+ *   redirectUri: "http://localhost:8000/callback",
+ *   scope: ["openid", Deno.env.get("AZURE_AD_CLIENT_ID")!]
+ * });
+ * ```
+ *
+ * @see {@link https://learn.microsoft.com/en-us/entra/identity-platform/v2-protocols-oidc}
+ */
+export function createAzureADOAuthConfig(config: {
```
Comment (iuioiua) [inline lib/create_azure_ad_oauth_config.ts:26]:
```suggestion
export function createAzureAdOAuthConfig(config: {
```
Nit: the Deno naming convention doesn't capitalize abbreviations.

**Review by iuioiua — CHANGES_REQUESTED** [review]

**Comment by iuioiua** [comment]
> > This is looking much better! Can you please provide ensure you've followed [these steps](https://github.com/denoland/deno_kv_oauth/blob/main/.github/CONTRIBUTING.md#adding-a-pre-configured-oauth-client) in this PR so we know the client functions are working?
> 
> 2 and 3 from that link are taken care of, i'm not sure what to do for 1 as the link on contributing is broken. I am using both configurations in a deployment, and they are working. Let me know what sort of screenshot i should be providing.

I've now corrected those links. Check it out now. It's about the screenshot of the demo working locally for you, so we know that the function works correctly. You may need to blur out any confidential text in this case.

**Comment by mcgear** [comment]
Date night tonight, but i should be able to get this worked up over the weekend so we can close this out.

File: lib/create_azure_adb2c_oauth_config.ts (hunk 5 above)
Comment (mcgear) [reply lib/create_azure_adb2c_oauth_config.ts]: fixed

File: lib/create_azure_ad_oauth_config.ts (hunk 6 above)
Comment (mcgear) [reply lib/create_azure_ad_oauth_config.ts]: fixed

**Comment by mcgear** [comment]
Alright, i think we are pretty close here:
AzureAdb2cSignedIn
AzureAdb2cSignedOut
AzureAdSignedIn
AzureAdSignedOut

**Review by iuioiua — APPROVED** [review]
LGTM! Excellent work. Thank you very much for this. It's a welcome addition.
